### ISSUE TITLE
`dart analyze` does not know when analyzer plugin results are complete

### ISSUE THREAD

**Issue by derolf** [dart-lang/sdk#38407]
`dartanalyzer version 2.5.0-edge.1ef83b86ae637ffe7359173804cbc6d3fa25e6db`

I have a plugin that works fine in the Analysis Server and Analysis notifcations show up in Visual Code.

If I run `dartanalyzer .`, I don't see notifications.

**Comment by bwilkerson** [comment]
Unfortunately we have not yet been able to work on getting plugins to work from the command-line analyzer. That work is desired, but not currently scheduled.

**Comment by zanderso** [comment]
Flutter's CLI users are asking for this to help in running analyzer plugins in their CI. See: https://github.com/flutter/flutter/issues/28327#issuecomment-689918445. /cc @devoncarew.

**Comment by bwilkerson** [comment]
Actually, they're asking for `flutter analyze` to run plugins, and my understanding is that `flutter analyze` uses the analysis server. I don't know why the analysis server is failing to run their plugins when run through `flutter analyze`, but it's a different issue than this one.

**Comment by zanderso** [comment]
Ah, sorry about that. I remember now that the issue with the analysis server is that it starts up plugins asynchronously, and doesn't give any indication when results from plugins are complete: https://github.com/flutter/flutter/issues/28327#issuecomment-532875896. Guessing that those problems would have to be solved in order to solve this one?

**Comment by bwilkerson** [comment]
Yes, thanks, that's probably the underlying issue here.

One possible solution would be to add a "command-line" mode for server that causes it to not report diagnostics until all of the plugins have reported results (with safeguards to protect against broken plugins, of course).

**Comment by devoncarew** [comment]
`dart analyze` is also using the analysis server's status `analysis` events to know when analysis starts and stops. That's not really what that event is intended to signify. We could:

- change the status.analysis event to fire isAnalyzing false until after the plugins report their results, and update the docs for the status message
- add a new event that clients could use to know when the analysis server was completely done processing a set of changes

**Comment by srawlins** [comment]
As `dart analyze` is the new primary supported tool, I've changed the title of this issue.

**Comment by bwilkerson** [comment]
We might want to close the issue. The `dartanalyzer` tool did not run plugins, but `dart analyze` does.

**Comment by bwilkerson** [comment]
Or, after briefly looking at the conversation, change the title to reflect the new problem: that there's no indication from server when all plugins have provided their results.

**Comment by incendial** [comment]
It'd be a great feature to have for Dart Code Metrics and sometimes we get random CI failings like this one: https://github.com/dart-code-checker/dart-code-metrics/runs/3163744741?check_suite_focus=true (Note: the issues are from Dart Code Metrics) or running [flutter analyze](https://github.com/dart-code-checker/dart-code-metrics/issues/322#issuecomment-836476417) which indicates that the server can actually handle external issues. Also, we receive similar reports from our users too.

We report issues to analyzer server using `channel.sendNotification`.

Can we help with this problem somehow?

**Comment by bwilkerson** [comment]
My understanding of how this works (it's been a while since I looked at the code) is that the server sends a `server.status` notification to the client when server has finished analyzing all of the files that it's going to analyze, and that `dart analyze` is using this as a signal that it can stop waiting for `analysis.errors` notifications and display the information that it has.

The `server.status` notification wasn't designed to be used for that purpose, and has only ever been intended as feedback to users of IDEs that enough analysis had been completed that they (the users) could expect complete results from various features, such as search.

What's required is a more reliable signal from the server that there will be no further `analysis.errors` notifications unless the content of the files being analyzed changes. Whether we repurpose the `server.status` notification or create a new notification is a design question [1]. The bigger issue, in my mind, is that there is currently no signal from a plugin to the server to indicate that the plugin will not be sending any additional `analysis.errors` notifications. I don't think we want to make the assumption that all plugins will always analyze every `.dart` file, nor that they will only analyze `.dart` files, so I don't think there's currently any way for server to know when plugins are done.

Once that's done I suspect that it won't be too hard to keep track of the number of running plugins that have not signaled full diagnostics since the last `analysis.handleWatchEvents`, `analysis.setContextRoots`, or `analysis.updateContent` request, and for server to wait to signal full diagnostics to the command-line tool until all the plugins have also completed their analysis.

[1] To be clear, I'd be fine with re-using the existing notification; I think the new semantics are consistent with the current semantics, but it's worth asking whether anyone knows of any reason why that would be a poor choice.

**Comment by incendial** [comment]
@bwilkerson If I understood your comment before and the code correctly, we need to change `isAnalyzing` getter to also rely on the plugins status (thus not changing the whole logic with `server.status`).

https://github.com/dart-lang/sdk/blob/main/pkg/analyzer/lib/src/dart/analysis/driver.dart#L1938
<img width="478" alt="Screenshot 2022-10-19 at 21 17 00" src="https://user-images.githubusercontent.com/21116742/196760083-3ebbcf93-4872-4347-a4f4-33ca98f448ae.png">

<img width="425" alt="Screenshot 2022-10-19 at 21 17 50" src="https://user-images.githubusercontent.com/21116742/196760238-b1a04f78-faad-4285-8115-12656551e079.png">

`AnalysisDriverScheduler` has access to the plugins via `DriverWatcher` (which is `PluginWatcher`). So, if the plugin watcher knows the current state of plugin analysis - the whole flow will work as expected?

If we're talking about the `dart analyze`, it means that we're not interested in the changes applied to the files after the command was started. But does this mean that the plugin will get all the context roots info at its creation?

Assuming all above is correct, what do you think about simply creating 2 new events: the first one indicating that the plugin will contribute to the errors list and the second indicating the end of the analysis?

I'm still not sure how to properly calculate the plugin progress. For DCM use-case just collecting the list of dart files from `AnalysisContextCollection.analysisContext.contextRoot.analyzedFiles()` (doing so for each context) and then on `analyzeFile` just marking each file as analyzed until there are zero files looks enough, but maybe I'm missing something?

Could you please share your thoughts?

**Comment by bwilkerson** [comment]
> ... we need to change `isAnalyzing` getter to also rely on the plugins status (thus not changing the whole logic with `server.status`).

I haven't looked at the code recently, but assuming that `isAnalyzing` is used to determine what status to return via `server.status`, then yes, that's at least one possible implementation path.

> So, if the plugin watcher knows the current state of plugin analysis - the whole flow will work as expected?

I believe that's the case, but as I said it's been a while since I looked at the code, so no guarantees.

> If we're talking about the `dart analyze`, it means that we're not interested in the changes applied to the files after the command was started. But does this mean that the plugin will get all the context roots info at its creation?

The `dart analyze` command is a thin wrapper over the analysis server. It passes the files and directories from the command line to the `analysis.setAnalysisRoots` request, then watches for `analysis.errors` notifications until it has seen a `server.status` notification indicating that analysis has finished. It then prints the diagnostics that it has received and exits. So yes, the plugins should receive all of the context roots as part of handling the `analysis.setAnalysisRoots` request just like it would if it were running in an IDE.

> ... what do you think about simply creating 2 new events: the first one indicating that the plugin will contribute to the errors list and the second indicating the end of the analysis?

I think there's already an assumption that plugins will contribute diagnostics, so the first event probably isn't necessary. The server will need some indication that it should expect to get status notifications, and a separate notification could serve that purpose, but so could sending a `plugin.status` notification indicating that analysis has started.

I think that those two pieces of information (server should wait for a "done analyzing" notification and a "done analyzing" notification) would solve the immediate problem. Unfortunately it would break one of the design principles, which is that a poorly behaving plugin shouldn't be allowed to degrade the user experience. To satisfy that criteria we'd probably also need to have a time-out. If a plugin said it had started to analyze code but hasn't indicated that analysis is complete, and if it has been more than N ms since the last `analysis.errors` notification from the plugin, then server should assume that the plugin is broken and tell the client that it's done with analysis.

To have a better UX we'd probably want to expand the `server.status` notification to allow a list of the names of plugins that didn't complete to be returned so that `dart analyze` could tell the user that the results might be incomplete.

> ... just marking each file as analyzed until there are zero files ...

If the status is a simple "we're done", then counting files shouldn't be necessary.

**Comment by incendial** [comment]
Thank you for you detailed comment!

I'm working on the implementation and looks like my initial suggestion about the plugin watcher won't work since `AnalysisDriverScheduler` doesn't know anything about it (only about generic DriverWatcher) and seems that it's intentional and shouldn't be changed. So, I'll try to merge `AnalysisDriverScheduler` and `PluginManager` results in `LegacyAnalysisServer`, looks like this should work. This and creating new notification for plugin analysis completion and I'll be able to land the draft CL for more direct discussion.

If you have any better suggestions, I'd really appreciate that 🙂 

> Unfortunately it would break one of the design principles, which is that a poorly behaving plugin shouldn't be allowed to degrade the user experience. To satisfy that criteria we'd probably also need to have a time-out. If a plugin said it had started to analyze code but hasn't indicated that analysis is complete, and if it has been more than N ms since the last analysis.errors notification from the plugin, then server should assume that the plugin is broken and tell the client that it's done with analysis.

Great point! I've added a timer.

> If the status is a simple "we're done", then counting files shouldn't be necessary.

Sorry, I think I messed up the explanation again 😄. I assume that we only want to send the "done analysing" notification from the plugin only once. So, the problem from the plugin side is to properly understand when this "done" is happened (for instance, DCM overrides `analyzerFile`, but we need to somehow understand that all the available files were analyzed). That's why I mentioned this idea of counting files.

Btw, looks like plugin path here https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/plugin/plugin_watcher.dart#L69 and there https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/plugin/plugin_manager.dart#L525 are different values, but they both are used as keys in `_pluginMap`

**Comment by rrousselGit** [comment]
On a different topic: Could a quick fix be to not start plugins at all when using `dart analyze`?

Because currently the output of `dart analyze` is subject to a race condition. If a plugin starts quickly enough and emits a lint, then it'll show up in the output.
But if it's a bit slower, then it won't.

That makes its behavior unpredictable.

And considering current plugins expose a custom command line to list lints specific to said plugin, it's probably reasonable to just not start the plugins at first (until the work to support this for real is done).

**Comment by incendial** [comment]
If I'm not mistaken, the `LegacyAnalysisServer` is used for `dart analyze` and `LSPAnalysisServer` is used for the IDE integration. If this is true, then passing a flag to the `PluginManager` not to add plugins at all will do the trick.

**Comment by bwilkerson** [comment]
> If I'm not mistaken, the `LegacyAnalysisServer` is used for `dart analyze` ...

That's true. It's also used for IntelliJ and Android Studio support.

If we want to disable plugins for the `dart analyze` case we'll probably have to depend on the client id passed to the server when it's starting.

**Comment by incendial** [comment]
>  It's also used for IntelliJ and Android Studio support.

Oh, that was unexpected, I though IDEA support LSP too

**Comment by bwilkerson** [comment]
They have a new product named `Fleet` that supports LSP, but to the best of my knowledge IntelliJ doesn't. I'm not aware of any LSP support for Android Studio either.

**Comment by incendial** [comment]
@bwilkerson I think I managed to make it work, but the overall analysis now takes **a lot** more time 😞 (like 15 minutes instead of 20-40 seconds), and the warnings from the analyzer itself are duplicated. I'm testing on DCM and not sure what's causing this, since in the IDE or our own CLI everything is much faster. 
Initially I've added the timer, but commented it out for now, looks like the performance problem should be addressed first.
<img width="1073" alt="Screenshot 2022-10-21 at 16 12 59" src="https://user-images.githubusercontent.com/21116742/197193042-5866cee5-9e2b-47eb-8130-494c41d17f5f.png">

And I'm getting a strange crash while running `dart analyze`:
<img width="765" alt="Screenshot 2022-10-21 at 16 15 23" src="https://user-images.githubusercontent.com/21116742/197193440-be8bc604-07f2-4c4b-951f-82ad57dccd82.png">

Also noticed, that https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/plugin/plugin_manager.dart#L922 is not used anywhere.

**Comment by incendial** [comment]
@bwilkerson could you please take a look https://dart-review.googlesource.com/c/sdk/+/265089? I was testing it with https://github.com/dart-code-checker/dart-code-metrics/pull/1048

**Comment by bwilkerson** [comment]
I'm sorry, but I don't have any ideas as to why such a change would cause it to take so much longer. My first guess would be that server is waiting for a signal from the plugin and is either not receiving it or is ignoring it. But I suspect that you've already checked for those conditions.

The duplicated diagnostics could easily be a bug in `dart analyze`. It was probably written with the assumption that the server would send a single notification for each file (which would presumably be true in the absence of plugins), and hence isn't dealing with the need to replace earlier lists of diagnostics with the updated lists when a second notification is received. The scenario I'm envisioning is that server analyzes the file, sending those diagnostics to the client (`dart analyze` in this case) and caching them. When it received diagnostics from the plugin it adds them to the cached list and then sends the combined list to the client. If the client doesn't discard the previous list then it could end up with duplicated diagnostics.

I also don't know why `isNonResponsive` isn't being used. It certainly seems reasonable to me that server would want to be able to restart plugins that are no longer responding, but I can only guess that either we found a different way to do that or that we never finished that arc of work. As I've said before, this is just a proof-of-concept, not production ready code.

**Comment by incendial** [comment]
> My first guess would be that server is waiting for a signal from the plugin and is either not receiving it or is ignoring it.

The plugin is the problem, but I was not able to find out why. I've seen that if I send the "analysis finished" event from the manager just after the first set of errors - it took like 10 extra seconds, but if I wait for the analysis_finished notification - it takes that much time.

**Comment by bwilkerson** [comment]
I looked at the CL and made a couple of comments. Not sure whether they point to the cause of the issue you're seeing or not.

**Comment by Toglefritz** [comment]
Hello @bwilkerson,

I spent some time investigating this issue while working on Dart analyzer plugins and wanted to share what I found.

I created a minimal reproduction that demonstrates a Dart project using an analyzer plugin: [https://github.com/Toglefritz/dart_analyzer_plugin_test/tree/main/minimal_dart_example](https://github.com/Toglefritz/dart_analyzer_plugin_test/tree/main/minimal_dart_example)

I tested this project using two different SDKs.

## Dart 3.12.2 (stable)

First, I tested using Dart 3.12.2, which is the version bundled with the current stable Flutter SDK at the time of writing.

```
$ dart --version
Dart SDK version: 3.12.2 (stable)
```

Running dart analyze causes the analysis server to crash:

```
Analyzing minimal_dart_example...
/Bad state: The analysis server crashed unexpectedly
```

The process then stalls indefinitely and never completes.

As a control experiment, I ran the same project with plugins disabled with `dart analyze --no-plugins`. This completes successfully:

```
Analyzing minimal_dart_example... 1.4s
No issues found!
```

This suggests the hang is specifically related to analyzer plugin support rather than `dart analyze` itself.

## Current main branch

Next, I built the Dart SDK locally from the current main branch and repeated the same test.

```
$ dart --version
Dart SDK version: 3.14.0-edge...
```

With the locally built SDK, dart analyze completes successfully and reports diagnostics from the analyzer plugin as expected.

I also added some temporary instrumentation to the analysis server to observe the plugin analysis lifecycle. The logs show the plugin analysis beginning, the analysis status transitioning correctly, and finally the server reporting that analysis is complete before the CLI exits. The custom diagnostics are then printed as expected.

The output is:

```
info • example/trigger_lint.dart:5:15 • The integer literal 42 is forbidden. Use a different number. • ban_42
info • lib/src/ban_42_rule.dart:38:23 • The integer literal 42 is forbidden. Use a different number. • ban_42

2 issues found.
```

## Conclusion

Based on these results, I believe the behavior described in this issue has already been fixed somewhere between Dart 3.12.2 and the current main branch.

The current SDK correctly waits for plugin analysis to complete and reports plugin diagnostics through `dart analyze` in my testing. The failure I observe appears to be specific to the older stable SDK, where the analysis server crashes when plugins are enabled.

If this issue is intended to track support for analyzer plugins in `dart analyze`, it appears to be resolved on the current HEAD of the Dart SDK. It may simply need verification against a newer SDK release (likely Dart 3.13 or later) before being closed.

**Comment by bwilkerson** [comment]
@srawlins Can you confirm that the issue has been addressed?

**Comment by srawlins** [comment]
@Toglefritz sorry for the issue you encountered; this issue is for the legacy plugin system. Your issue, I believe, is with the new plugin system. And I believe it is solved in Dart 3.14. If not, please file a new issue and we can track that.

We've deprecated the legacy plugin system, as of Dart 3.14.2, and will be removing support for the legacy plugin system in an upcoming release of the Dart SDK. The [new plugins system](https://dart.dev/tools/analyzer-plugins) is supported and continually improving.
